### PR TITLE
HPCC-26227 React version of Package Maps page

### PR DESCRIPTION
**Pull request by jeclrsg**
Signed-off-by: Jeremy Clements <jeremy.clements@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/src/src-react/components/PackageMapPartDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    part
+}) => {
+
+    const [_package, setPackage] = React.useState<any>(undefined);
+    const [xml, setXml] = React.useState("");
+
+    React.useEffect(() => {
+        WsPackageMaps.PackageMapQuery({}).then(({ ListPackagesResponse }) => {
+            const __package = ListPackagesResponse?.PackageMapList?.PackageListMapData.filter(item => item.Id === name)[0];
+            console.log(ListPackagesResponse);
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMapPartDetails.tsx:21]: Remove console.log?

File: esp/src/src-react/components/PackageMapParts.tsx [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                    });
+                }
+            }
+        },
+    ], [_package, gridStore, name, refreshTable, selection, uiState.hasSelection]);
+
+    React.useEffect(() => {
+        WsPackageMaps.getPackageMapById({ packageMap: name }).then(({ GetPackageMapByIdResponse }) => {
+            const xml = parser.parse(GetPackageMapByIdResponse?.Info);
+            const parts = [...xml.getElementsByTagName("Part")].map(part => {
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMapParts.tsx:111]: Not sure the outer `[... ]` adds anything given its about to be mapped anyway?

File: esp/src/src-react/components/PackageMapParts.tsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        });
+    }, [name]);
+
+    //  Selection  ---
+    React.useEffect(() => {
+        const state = { ...defaultUIState };
+
+        for (let i = 0; i < selection.length; ++i) {
+            state.hasSelection = true;
+            //  TODO:  More State
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMapParts.tsx:134]: TODO?

File: esp/src/src-react/components/PackageMaps.tsx [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+            key: "refresh", text: nlsHPCC.Refresh, iconProps: { iconName: "Refresh" },
+            onClick: () => refreshTable()
+        },
+        { key: "divider_1", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
+        {
+            key: "open", text: nlsHPCC.Open, disabled: !uiState.hasSelection,
+            onClick: () => {
+                selection.forEach(item => {
+                    window.open(`#/packagemaps/${item?.Id}`, "_blank");
+                });
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMaps.tsx:219]: Typically only open in a new tab if selection.length > 1?

File: esp/src/src-react/components/PackageMaps.tsx [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+        refreshTable();
+    }, [filter, refreshTable]);
+
+    //  Selection  ---
+    React.useEffect(() => {
+        const state = { ...defaultUIState };
+
+        for (let i = 0; i < selection.length; ++i) {
+            state.hasSelection = true;
+            //  TODO:  More State
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMaps.tsx:346]: TODO?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/src-react/components/PackageMapParts.tsx (hunk 2 above)
Comment (jeclrsg) [reply esp/src/src-react/components/PackageMapParts.tsx:111]: Needed to be converted before calling map(), getElementsByTagName returns an HTMLCollection which is array-like, but doesn't have most of the functions of Array.prototype

File: esp/src/src-react/components/PackageMapPartDetails.tsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+}) => {
+
+    const [_package, setPackage] = React.useState<any>(undefined);
+    const [xml, setXml] = React.useState("");
+
+    React.useEffect(() => {
+        WsPackageMaps.PackageMapQuery({}).then(({ ListPackagesResponse }) => {
+            const __package = ListPackagesResponse?.PackageMapList?.PackageListMapData.filter(item => item.Id === name)[0];
+            setPackage(__package);
+        });
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMapPartDetails.tsx:22]: logger + catch + error reporting

File: esp/src/src-react/components/PackageMapPartDetails.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        if (!_package || !name || !part) return;
+        WsPackageMaps.GetPartFromPackageMap({
+            request: {
+                PackageMap: name.split("::")[1],
+                Target: _package?.Target,
+                PartName: part
+            }
+        }).then(({ GetPartFromPackageMapResponse, Exceptions }) => {
+            setXml(GetPartFromPackageMapResponse?.Content);
+        });
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMapPartDetails.tsx:35]: logger + catch + error reporting

File: esp/src/src-react/components/PackageMaps.tsx [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+                Info: activeMapXml
+            }
+        }).then(({ValidatePackageResponse, Exceptions}) => {
+            if (Exceptions?.Exception.length > 0) {
+                setShowError(true);
+                setErrorMessage(Exceptions?.Exception[0].Message);
+            } else {
+                setActiveMapValidationResult(validateResponseToText(ValidatePackageResponse));
+            }
+        });
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMaps.tsx:145]: logger + catch + error reporting

File: esp/src/src-react/components/PackageMaps.tsx [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+                Info: contentsXml
+            }
+        }).then(({ValidatePackageResponse, Exceptions}) => {
+            if (Exceptions?.Exception.length > 0) {
+                setShowError(true);
+                setErrorMessage(Exceptions?.Exception[0].Message);
+            } else {
+                setContentsValidationResult(validateResponseToText(ValidatePackageResponse));
+            }
+        });
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMaps.tsx:162]: logger + catch + error reporting

File: esp/src/src-react/components/PackageMaps.tsx [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+    }, [contentsTarget, contentsXml, setContentsValidationResult, setErrorMessage, setShowError]);
+
+    React.useEffect(() => {
+        if (activeMapProcess !== "" && activeMapTarget !== "") {
+            WsPackageMaps.getPackage({ target: activeMapTarget, process: activeMapProcess }).then(({GetPackageResponse}) => {
+                setActiveMapXml(GetPackageResponse.Info);
+            });
```
Comment (GordonSmith) [inline esp/src/src-react/components/PackageMaps.tsx:169]: logger + catch + error reporting

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**